### PR TITLE
Capture Nextflow logs via syslog rather than console

### LINKED ISSUE #54: ANSI logging
By default nextflow uses an asynchronous logging to console output, which is not supported by nftest, which makes the output very messy. We should turn it off using `-ansi-log` from nextflow when logging into the `log-nftest-${date}.log` file, and we should also turn it off for the console output, unless we want to somehow let nftest capture the asynchronous logging and also print out asynchronously.

See here: https://www.nextflow.io/docs/latest/cli.html#options

### PR DESCRIPTION
**Pull request by nwiltsie**
# Description
This change uses Nextflow's `-syslog` [command line flag](https://www.nextflow.io/docs/edge/cli.html#logging-to-a-syslog-server) to capture more structured logs than from the console. (The console logs are also suppressed with `-quiet`.) Lines printed to the console are still logged at `INFO`-level to a new `console` logger.

The active thread for each log message is now recorded to the logfile. It is not displayed on the console for simplicity.

```
### console
2024-01-17 20:32:03,951 - NFTest - INFO - Beginning tests...
2024-01-17 20:32:03,952 - NFTest - INFO - A-mini-n2: None

### logfile
2024-01-17 20:32:03,951 - NFTest - MainThread - INFO - Beginning tests...
2024-01-17 20:32:03,952 - NFTest - MainThread - INFO - A-mini-n2: None
```

I also fixed a bug I introduced with #58 where Nextflow logs go to `NFTest` rather than `nextflow`.

Despite the additional complexity there are several advantages to this approach. First, Nextflow only posts "relevant" process statuses, as opposed to the periodic status updates noted in #54:

Before:
```
2024-01-17 20:31:07,041 - NFTest - INFO - executor >  local (8)
2024-01-17 20:31:07,041 - NFTest - INFO - [72/6536cb] process > run_validate_PipeVal (4)       [100%] 4 of 4 ✔
2024-01-17 20:31:07,041 - NFTest - INFO - [03/b45ace] process > run_selectPositions_REDItools  [100%] 1 of 1 ✔
2024-01-17 20:31:07,041 - NFTest - INFO - [36/cefb0e] process > run_AnnotateRepeatMasker_RE... [100%] 1 of 1 ✔
2024-01-17 20:31:07,042 - NFTest - INFO - [93/c89093] process > run_AnnotateGene_REDItools     [100%] 1 of 1 ✔
2024-01-17 20:31:07,042 - NFTest - INFO - [37/57af12] process > run_FilterREDIportal_REDItools [  0%] 0 of 1
2024-01-17 20:31:07,042 - NFTest - INFO -
2024-01-17 20:31:07,042 - NFTest - INFO -
2024-01-17 20:31:07,042 - NFTest - INFO -
2024-01-17 20:31:07,042 - NFTest - INFO -
2024-01-17 20:31:07,042 - NFTest - INFO - Script time --> START: 17/01/2024 20:30:42
2024-01-17 20:31:07,043 - NFTest - INFO - Reading table...
2024-01-17 20:31:07,043 - NFTest - INFO - Total lines: 653
2024-01-17 20:31:07,043 - NFTest - INFO - Filtered in lines: 74
2024-01-17 20:31:07,043 - NFTest - INFO - Selected lines saved on CPCG0196-F1-A-mini-n2_candidates.txt
2024-01-17 20:31:07,043 - NFTest - INFO - Script time --> END: 17/01/2024 20:30:42
2024-01-17 20:31:07,043 - NFTest - INFO -
2024-01-17 20:31:07,043 - NFTest - INFO -
2024-01-17 20:31:07,043 - NFTest - INFO -
2024-01-17 20:31:07,044 - NFTest - INFO -
2024-01-17 20:31:07,115 - NFTest - INFO - executor >  local (8)
2024-01-17 20:31:07,116 - NFTest - INFO - [72/6536cb] process > run_validate_PipeVal (4)       [100%] 4 of 4 ✔
2024-01-17 20:31:07,116 - NFTest - INFO - [03/b45ace] process > run_selectPositions_REDItools  [100%] 1 of 1 ✔
2024-01-17 20:31:07,116 - NFTest - INFO - [36/cefb0e] process > run_AnnotateRepeatMasker_RE... [100%] 1 of 1 ✔
2024-01-17 20:31:07,116 - NFTest - INFO - [93/c89093] process > run_AnnotateGene_REDItools     [100%] 1 of 1 ✔
2024-01-17 20:31:07,116 - NFTest - INFO - [37/57af12] process > run_FilterREDIportal_REDItools [100%] 1 of 1 ✔
```

After:
```
2024-01-17 20:32:22,326 - nextflow - INFO - [01/706655] Submitted process > run_selectPositions_REDItools
2024-01-17 20:32:23,131 - console - INFO - Script time --> START: 17/01/2024 20:32:22
2024-01-17 20:32:23,179 - nextflow - INFO - [0f/1a7d38] Submitted process > run_AnnotateRepeatMasker_REDItools
2024-01-17 20:32:27,876 - nextflow - INFO - [bb/c97103] Submitted process > run_AnnotateGene_REDItools
2024-01-17 20:32:34,434 - nextflow - INFO - [39/0a83e3] Submitted process > run_FilterREDIportal_REDItools
2024-01-17 20:32:43,921 - console - INFO - Reading table...
2024-01-17 20:32:43,921 - console - INFO - Total lines: 653
2024-01-17 20:32:43,921 - console - INFO - Filtered in lines: 74
2024-01-17 20:32:43,921 - console - INFO - Selected lines saved on CPCG0196-F1-A-mini-n2_candidates.txt
2024-01-17 20:32:43,922 - console - INFO - Script time --> END: 17/01/2024 20:32:22
2024-01-17 20:32:43,922 - console - INFO - All positions: 74
2024-01-17 20:32:46,349 - console - INFO - Positions filtered in: 0
```

Second, Nextflow pushes `DEBUG`-level logs, and logs from more modules, to the syslog server. Most of these logs would otherwise only be saved in the `.nextflow.log` file and be very difficult to merge with the NFTest logs:

Before:
```
2024-01-17 20:30:26,970 - NFTest - INFO - N E X T F L O W  ~  version 23.04.2
2024-01-17 20:30:31,855 - NFTest - INFO - Launching `./main.nf` [sick_carson] DSL2 - revision: 7ffe311c28
```

After:
```
2024-01-17 20:32:06,930 - nextflow - main - INFO - N E X T F L O W  ~  version 23.04.2
2024-01-17 20:32:06,954 - nextflow - main - DEBUG - Setting up plugin manager > mode=prod; embedded=false; plugins-dir=/mnt/exports/shared/home/nwiltsie/.nextflow/plugins; core-plugins: nf-amazon@1.16.2,nf-azure@1.0.1,nf-codecommit@0.1.4,nf-console@1.0.5,nf-ga4gh@1.0.5,nf-google@1.7.3,nf-tower@1.5.12,nf-wave@0.8.3
2024-01-17 20:32:07,032 - nextflow - main - INFO - Enabled plugins: []
2024-01-17 20:32:07,042 - nextflow - main - INFO - Disabled plugins: []
2024-01-17 20:32:07,043 - nextflow - main - INFO - PF4J version 3.4.1 in 'deployment' mode
2024-01-17 20:32:07,077 - nextflow - main - INFO - No plugins
2024-01-17 20:32:07,106 - nextflow - main - DEBUG - Found config local: /hot/code/nwiltsie/pipelines/pipeline-filter-RNAEditingSite/nextflow.config
2024-01-17 20:32:07,122 - nextflow - main - DEBUG - User config file: /hot/code/nwiltsie/pipelines/pipeline-filter-RNAEditingSite/test/nftest.config
2024-01-17 20:32:07,122 - nextflow - main - DEBUG - Parsing config file: /hot/code/nwiltsie/pipelines/pipeline-filter-RNAEditingSite/nextflow.config
2024-01-17 20:32:07,123 - nextflow - main - DEBUG - Parsing config file: /hot/code/nwiltsie/pipelines/pipeline-filter-RNAEditingSite/test/nftest.config
2024-01-17 20:32:07,206 - nextflow - main - DEBUG - Applying config profile: `standard`
2024-01-17 20:32:08,386 - nextflow - main - DEBUG - Applying config profile: `standard`
2024-01-17 20:32:11,695 - nextflow - main - DEBUG - Applied DSL=2 from script declararion
2024-01-17 20:32:11,727 - nextflow - main - INFO - Launching `./main.nf` [modest_kirch] DSL2 - revision: 7ffe311c28
```

---

## Nitty-gritty details

[Syslog](https://en.wikipedia.org/wiki/Syslog) is a standard message logging format that is [supported by Nextflow](https://www.nextflow.io/docs/edge/cli.html#logging-to-a-syslog-server).

In order to capture the logs, each test case now spins up a custom [`socketserver.UDPServer`](https://docs.python.org/3/library/socketserver.html#socketserver-udpserver-example) in a separate thread. The server unconditionally logs the raw syslog messages at INFO level to the `nextflow` logger (see below). Context managers make sure that the server (and thus the thread) are both torn down once the `nextflow run` command returns.

Nextflow uses the [BSD syslog format](https://datatracker.ietf.org/doc/html/rfc3164). In order to extract the embedded information from the raw syslog messages (logging level, thread, etc.) I'm using a [filter](https://docs.python.org/3/library/logging.html#filter-objects) attached to the `nextflow` logger.

### Problems

The one irritation is that most syslog implementations (including Nextflow's) are based on UDP and thus are vulnerable to message re-ordering and packet loss. As the server is on the same host and is using a [ThreadingUDPServer](https://docs.python.org/3/library/socketserver.html#socketserver.ThreadingUDPServer) that handles each connection in a new thread, I'm not particularly concerned about dropping packets due to congestion.

_However_, syslog messages are also limited to the size of 1/2 of the system's socket send buffer size (`SO_SNDBUF`, [docs](https://man7.org/linux/man-pages/man7/socket.7.html)), or 8192 bytes on my F2. When a workflow finishes Nextflow logs the entire trace object, which ends up getting truncated:

```
2024-01-17 20:32:44,199 - nextflow - main - DEBUG - Workflow completed -- rendering execution report
2024-01-17 20:32:44,338 - nextflow - main - DEBUG - Execution report summary data:
  [{"cpuUsage": ---snip--- "q2Label":"run_AnnotateGene_RED     <--- That's just the end, no closing quote
2024-01-17 20:32:45,688 - nextflow - main - DEBUG - Workflow completed -- rendering execution timeline
2024-01-17 20:32:46,117 - nextflow - main - DEBUG - Closing CacheDB done
```

The truncated message does not cause any other problems, but it's just incorrect and I don't like it.

## Pipeline Run Results

Clean results (before these changes):
* `/hot/software/pipeline/pipeline-filter-RNAEditingSite/Nextflow/development/unreleased/nwiltsie-syslog-testing/log-nftest-20240117T203023Z.log`
* `/hot/software/pipeline/pipeline-filter-RNAEditingSite/Nextflow/development/unreleased/nwiltsie-syslog-testing/sbatch-nftest-20240117T203018Z.log`

Results after these changes:
* `/hot/software/pipeline/pipeline-filter-RNAEditingSite/Nextflow/development/unreleased/nwiltsie-syslog-testing/log-nftest-20240117T203203Z.log`
* `/hot/software/pipeline/pipeline-filter-RNAEditingSite/Nextflow/development/unreleased/nwiltsie-syslog-testing/sbatch-nftest-20240117T203159Z.log`

# Checklist

- [x] This PR **does *NOT* contain** Protected Health Information [(PHI)](https://ohrpp.research.ucla.edu/hipaa/). A repo may ***need to be deleted*** if such data is uploaded. <br> Disclosing PHI is a ***major problem***[^1] - Even ***a small leak can be costly***[^2].
  
- [x] This PR **does *NOT* contain** germline genetic data[^3], RNA-Seq, DNA methylation, microbiome or other molecular data[^4].

[^1]: [UCLA Health reaches $7.5m settlement over 2015 breach of 4.5m patient records](https://healthitsecurity.com/news/ucla-health-reaches-7.5m-settlement-over-2015-breach-of-4.5m)
[^2]: [The average healthcare data breach costs $2.2 million, despite the majority of breaches releasing fewer than 500 records.](https://www.ponemon.org/local/upload/file/Sixth%20Annual%20Patient%20Privacy%20%26%20Data%20Security%20Report%20FINAL%206.pdf)
[^3]: [Genetic information is considered PHI.](https://www.genome.gov/about-genomics/policy-issues/Privacy#:~:text=In%202013%2C%20as%20required%20by,genetic%20information%20for%20underwriting%20purposes.)
  [Forensic assays can identify patients with as few as 21 SNPs](https://www.sciencedirect.com/science/article/pii/S1525157817305962)
[^4]: [RNA-Seq](https://www.nature.com/articles/ng.2248), [DNA methylation](https://ieeexplore.ieee.org/document/7958619), [microbiome](https://www.pnas.org/doi/pdf/10.1073/pnas.1423854112), or other molecular data can be used to predict genotypes (PHI) and reveal a patient's identity.

- [x] This PR **does *NOT* contain** other non-plain text files, such as: compressed files, images (*e.g.* `.png`, .`jpeg`), `.pdf`, `.RData`, `.xlsx`, `.doc`, `.ppt`, or other output files.

_&emsp; To automatically exclude such files using a [.gitignore](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files) file, see [here](https://github.com/uclahs-cds/template-base/blob/main/.gitignore) for example._

- [x] I have read the [code review guidelines](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3187646/Code+Review+Guidelines) and the [code review best practice on GitHub check-list](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List).

- [x] I have set up or verified the `main` branch protection rule following the [github standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3190380/GitHub+Standards#GitHubStandards-Branchprotectionrule) before opening this pull request.

- [x] The name of the branch is meaningful and well formatted following the [standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List), using [AD_username (or 5 letters of AD if AD is too long)]-[brief_description_of_branch].
  
- [x] I have added the major changes included in this pull request to the `CHANGELOG.md` under the next release version or unreleased, and updated the date.

### CODE REVIEW

**Comment by nwiltsie** [comment]
I changed the indent level of a big block of code, so it looks like I changed the content of more lines than I actually did. The "hide-whitespace" option helps: https://github.com/uclahs-cds/tool-NFTest/pull/62/files?diff=split&w=1

File: nftest/NFTestCase.py [hunk 1]
```diff
@@ -102,41 +108,69 @@ def test(self) -> bool:
...
-            sp.list2cmdline(nextflow_command)
-        )
-
-        process = popen_with_logger(
-            nextflow_command,
-            env={**os.environ, **envmod},
-            logger=self._logger
-        )
+        # Use ExitStack to handle the multiple nested context managers
+        with ExitStack() as stack:
```
Comment (yashpatel6) [inline nftest/NFTestCase.py:112]: Is the other context here the subprocess context? Does that need to be tied to the stack here or does it function independently?

File: nftest/syslog.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+LEVELS = [
+    logging.CRITICAL,   # 0 = Emergency
+    logging.CRITICAL,   # 1 = Alert
+    logging.CRITICAL,   # 2 = Critical
+    logging.ERROR,      # 3 = Error
+    logging.WARNING,    # 4 = Warning
+    logging.INFO,       # 5 = Notice
+    logging.INFO,       # 6 = Info
+    logging.DEBUG,      # 7 = Debug
+]
```
Comment (yashpatel6) [inline nftest/syslog.py:18]: Is this intended to map logging levels to another priority ranking?

File: nftest/syslog.py
Comment (yashpatel6) [inline nftest/syslog.py:1]: It may be worth adding some tests for the functions/classes here

**Review by yashpatel6 — COMMENTED** [review]
Generally looks good! A couple of comments:

File: nftest/syslog.py (hunk 2 above)
Comment (nwiltsie) [reply nftest/syslog.py:18]: It's mapping from [syslog severity levels](https://en.wikipedia.org/wiki/Syslog#Severity_level) _to_ logging levels. The syslog severity is embedded in the message, and I use it as the index into this array to get the appropriate python logging level.

File: nftest/syslog.py
Comment (nwiltsie) [reply nftest/syslog.py:1]: Yeah fair, I'll get some added.

File: nftest/syslog.py (hunk 2 above)
Comment (nwiltsie) [reply nftest/syslog.py:18]: Here's the relevant code chunk (and some more details about [priority](https://datatracker.ietf.org/doc/html/rfc5424#section-6.2.1)):

https://github.com/uclahs-cds/tool-NFTest/blob/f27ed68ee4263cd823d6da89493e334356c929c0/nftest/syslog.py#L42-L46

File: nftest/NFTestCase.py (hunk 1 above)
Comment (yashpatel6) [reply nftest/NFTestCase.py:112]: Ah got it, I'm fine with keeping it as is, just wanted to make sure I wasn't missing something

File: nftest/syslog.py
Comment (nwiltsie) [reply nftest/syslog.py:1]: Okay, I've added some tests for the `syslog_filter` method, which constitutes the bulk of the new logic here.

NFTest results after the changes:
`/hot/software/pipeline/pipeline-filter-RNAEditingSite/Nextflow/development/unreleased/nwiltsie-syslog-testing/log-nftest-20240118T221736Z.log`
`/hot/software/pipeline/pipeline-filter-RNAEditingSite/Nextflow/development/unreleased/nwiltsie-syslog-testing/sbatch-nftest-20240118T221725Z.log`

**Review by yashpatel6 — APPROVED** [review]
Looks good! Anything to add @zhuchcn ?

**Pull request merged**